contracts/StreamManager.vy

```diff
@@ -1,259 +1,464 @@
-    is for, this contract simply streamlines the creation and payment of those
-    defined goods or services.
+@dev The Stream contract is owned by `controller`, who is the recipient of all streams created by
+    this contract. The `controller` can specify any number of payment tokens that they can be
+    accepted for streaming. Anyone can create a stream targeting the `controller`, as long as it is
+    one of the tokens that `controller` has specified as accepting.
+
+    Streams can only be successfully created after passing through a set of payment term Validator
+    contracts. Validators can do any arbitrary logic on the parameters of an incoming stream, and
+    return a computed streaming rate that is aggregated across all the validators and becomes the
+    streaming rate that the Stream will vest funds to the `controller` at.
+
+    Anyone (not just the stream `owner`) can add more paid time to a Stream. Streams can be cancel-
+    led by the `owner` at any point after `MIN_STREAM_LIFE` has elapsed, which is a parameter de-
+    signed to enforce the amount it takes to provision the product or service being fascilitated by
+    this contract.
+
+    The purpose of these Streams is usually as an optimistic payment method for an off-chain good
+    or service, so the security properties of this contract reflect that. As such, in all extra-
+    ordinary situations, the `controller` should have the ability or right to create and enforce
+    the terms that the payment is for, this contract simply streamlines the creation and payment of
+    those defined goods or services.
 """
-
 from ethereum.ercs import IERC20
 
 from . import Validator
 
 
+# List of Validator contracts to check for Stream creation, funding, migrations
+# TODO: Use `Set` to ensure uniqueness, when available
 MAX_VALIDATORS: constant(uint8) = 10
 validators: public(DynArray[Validator, MAX_VALIDATORS])
 
-
-MAX_REASON_SIZE: constant(uint16) = 1024
 MIN_STREAM_LIFE: public(immutable(uint256))
-
+MAX_PRODUCTS: constant(uint8) = 20
 
 struct Stream:
+    owner: address
     token: IERC20
     amount_per_second: uint256
-    max_stream_life: uint256
     funded_amount: uint256
     start_time: uint256
-    last_pull: uint256
-    reason: Bytes[MAX_REASON_SIZE]
+    last_claim: uint256
+    products: DynArray[bytes32, MAX_PRODUCTS]
 
-num_streams: public(HashMap[address, uint256])
-streams: public(HashMap[address, HashMap[uint256, Stream]])
 
-
-owner: public(address)
 token_is_accepted: public(HashMap[IERC20, bool])
 
+# Global index of Streams
+num_streams: public(uint256)
+streams: public(HashMap[uint256, Stream])
+
+# Service Provider (has all Capabilities, also beneficiary of funding)
+controller: public(address)
+new_controller: public(address)
+new_controller_proposed: public(uint256)
+CONTROLLER_ACCEPTANCE_DELAY: constant(uint256) = 7 * 24 * 60 * 60  # 1 week
+
+event NewControllerProposed:
+    old: indexed(address)
+    new: indexed(address)
```

Comment:
did this, and also for the stream owner update event, and also made all the stream events use `stream_id` instead of just `id` (it's clearer)

5d17c1b
8cb68c9
dbdd779